### ISSUE TITLE
Create a middleware to normalize object attributes

### ISSUE THREAD

**Issue by elliottwilliams** [propershark/shark#23]
For Citybus, station names annoyingly have the [stop code included in them](https://cloud.githubusercontent.com/assets/910198/17839639/c924b4ac-67bb-11e6-8a9e-bd17b1218967.png). Similarly, some routes have[ schedule info hand-typed into them](https://cloud.githubusercontent.com/assets/910198/17839693/e37c2e64-67bd-11e6-95cc-23888ba83807.png).

I'd like to have a generic way to normalize object attributes, probably one configurable on a per-agency basis. I could see it integrating into an agency configuration somewhat like this (please ignore any un-idiomatic ruby):

``` ruby
# in config/citybus.rb
agency.use_middleware Normalize do |object|
  case object
  when Shark::Station
    object.name.sub! /(.*) - .*/, '\1'
  end
end
```

How does this sound to you? Do you think it should actually happen on the client side?

**Comment by faultyserver** [comment]
I think this is a reasonable suggestion for the server to handle. Otherwise, it's just data duplication. The configuration will probably end up in it's own `config/normalizer.rb` (-r added for convention) to avoid bloating the agency config itself.

Fun fact: CityBus has at least 4 different syntaxes that it uses when including stop codes in station names. I've seen " - ", ": ", "-" and no punctuation at all.

Luckily, the code for removing them is relatively simple, something like this ([explanation](https://regexper.com/#%2F%5Cs*%5CW*%5Cs*%28BUS%7CTEMP%29%5Cd%2B%5Cs*%24%2F))

``` ruby
object.name.sub!(/\s*\W*\s*(BUS|TEMP)\d+\s*$/, '')
```

Shedding the schedule information off of route names might be a little more difficult, depending on the consistency of their syntax (`/\(Begins[^\)]+\)/` might be good enough), but it shouldn't be impossibly hard.

Also, your code is pretty idiomatic, assuming there would be more than just `when Shark::Station` in the `case` statement.

**Comment by elliottwilliams** [comment]
Great! Probably okay to put this to the side for now, but let's plan on doing it before 1.0

(regexper looks fantastic—thanks for the link)

**Comment by faultyserver** [comment]
As it stands, the block that `use_middleware` takes is reserved as a configurator, so implementing configurable normalization won't be _quite_ that clean.

Instead, something like this will probably suffice:

``` ruby
agency.use_middleware Normalizer do |config|
  config.on_shark_station = Proc.new do |station|
    station.name.sub!(/\s*\W*\s*(BUS|TEMP)\d+\s*$/, '')
  end

  config.on_shark_route = Proc.new do |route|
    route.name.sub!(/\(Begins[^\)]+\)/, '')
  end

  ...
end
```

While this isn't quite as clean or immediately semantic, it does have the advantage of avoiding the `case...when` statement (a precedent for anything exposed as "configurable") and somewhat better separation of type handling.

**Comment by faultyserver** [comment]
Actually, having `Normalizer` configuration in `config/citybus.rb` doesn't particularly make sense, nor does it follow the precedent set by `Transport`. It should have it's own configuration file - `config/normalizer.rb`, and the line in the agency configuration should simply be

``` ruby
agency.use_middleware Normalizer
```

Since that precedent exists, I think it makes sense to consider removing the reserved configurator block from `use_middleware` and instead passing it directly to the middleware itself, but that's for another issue.

**Comment by faultyserver** [comment]
I'm going to mark this closed as of f7b8182d8c8c68270138fb0c5ce3df6cd3e775ea. The functionality appears to work as described here, and the configuration is giving desired results in my testing.
